README.md
<h1 align="center">
  <br>
  <img src="./src/images/greymatter-logo.png" alt="greymatter.io Logo" width="50">
  <br>
  greymatter.io UI Component Library
  <br>
</h1>

<h4 align="center">React components for greymatter.io user interfaces</h4>

<p align="center">
  <a href="https://github.com/prettier/prettier" target="_blank">
    <img src="https://img.shields.io/badge/formatted_with-prettier-ff69b4.svg"
         alt="styled with prettier">
  </a>
  <a href="https://www.styled-components.com/" target="_blank"><img src="https://img.shields.io/badge/styled_with-%F0%9F%92%85%20styled--components-orange.svg?colorB=daa357" alt="styled with styled-components"></a>
  <a href="https://buildkite.com/greymatter/gm-ui-components" target="_blank"><img src="https://badge.buildkite.com/a3ec6b66f6c5cfb9875c88175a351156dc728ac7a72f47dddb.svg" alt="tested on Buildkite"></a>
  <a href="https://opensource.org/licenses/mit-license.php" target="_blank"><img src="https://badges.frapsoft.com/os/mit/mit.svg?v=103" alt="licensed under MIT"></a>
</p>

<p align="center">
  <a href="#installation">Installation</a> •
  <a href="#usage">Usage</a> •
  <a href="#documentation">Documentation</a> •
  <a href="#compatability">Compatability</a> •
  <a href="#license">License</a>
</p>

# Installation

gm-ui-components is available as an npm package. It has 6 peer dependencies:

```sh
npm install --save gm-ui-components react react-dom styled-components inter-ui react-popper @popperjs/core
```

# Usage

An example using the Button component:

```javascript
import React from 'react';
import { render } from 'react-dom';
import { Button } from 'gm-ui-components';

function App() {
  return (
    <Button
        label="Hello World!"
        type="default"
        size="xl"
        active={true}
    />
  );
}

render(<App />, document.querySelector('#app'));
```

## Theming

The component library provides 2 themes out of the box - `keen` and `keenDark`. All components use the `keen` theme by default. 

To use `keenDark` or a custom theme, wrap your app in a `ThemeProvider` and pass the theme object:

```javascript
// App.js
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'react-dom';
import { Button, keenDark } from 'gm-ui-components';

function Button() {
  return (
    <Button
        label="Hello World!"
        type="default"
        size="xl"
        active={true}
    />
  );
}

render(
  <ThemeProvider theme={keenDark}>
    <App />
  </ThemeProvider>, 
  document.querySelector('#app'));
```

To extend the theme, you must pass a theme object that follows our schema. Here is an example of the schema and its defaults:

```javascript
{
  name: "Keen",

  COLOR_BRAND_PRIMARY: "#00b42b",
  COLOR_BRAND_SECONDARY: "#00b42b",

  COLOR_BACKGROUND_DEFAULT: "hsl(0,0%,100%)",
  COLOR_BACKGROUND_TWO: "hsl(0,0%,97%)",
  COLOR_BACKGROUND_THREE: "hsl(0,0%,94%)",

  COLOR_CONTENT_DEFAULT: "hsla(0,0%,0%,0.85)",
  COLOR_CONTENT_CONTRAST: "hsla(0,0%,0%,1)",
  COLOR_CONTENT_MUTED: "hsla(0,0%,0%,0.6)",
  COLOR_CONTENT_NONESSENTIAL: "hsla(0,0%,0%,0.3)",

  COLOR_INTENT_HIGHLIGHT: "#00b42b",
  COLOR_INTENT_SUCCESS: "#00b42b",
  COLOR_INTENT_DANGER: "#D83D22", // WCAG AA+
  COLOR_INTENT_WARNING: "#F7CD45", // WCAG AA+
  COLOR_INTENT_INFO: "#1E6DF6", // WCAG AA+

  COLOR_KEYLINE_DEFAULT: "hsla(0,0%,0%,8%)",
  COLOR_KEYLINE_SOLID: "hsl(0,0%,8%)",

  OPACITY_FULL: "1",
  OPACITY_LIGHT: "0.7",
  OPACITY_LIGHTER: "0.5",
  OPACITY_LIGHTEST: "0.15",

  // LAYOUT
  SPACING_BASE: 8,

  ZINDEX_ABYSS: "-9999",
  ZINDEX_FLOOR: "1",
  ZINDEX_DROPDOWN: "1010",
  ZINDEX_STICKY: "1020",
  ZINDEX_FIXED: "1030",
  ZINDEX_SCRIM: "1040",
  ZINDEX_MODAL: "1050",
  ZINDEX_POPOVER: "1060",
  ZINDEX_TOOLTIP: "1070",

  CORNER_RADIUS_SHARP: "2px", // spacingScale(0.5)
  CORNER_RADIUS_INPUT: "4px", // spacingScale(1)
  CORNER_RADIUS_CARD_SM: "4px",
  CORNER_RADIUS_CARD_DEFAULT: "6px",
  CORNER_RADIUS_CARD_LG: "8px",
  CORNER_RADIUS_MAX: "90000px",

  // TYPOGRAPHY
  FONT_STACK_DEFAULT: `-apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif`,
  FONT_STACK_BRAND: `"Metropolis", "Helvetica Neue", Arial, sans-serif`,
  FONT_STACK_CODE: `"SFMono-Regular", Consolas, "Liberation Mono", Menlo, Courier, monospace`,

  FONT_SIZE_PAGE_TITLE: "40px",

  FONT_SIZE_HEADING_LG: "38px",
  FONT_SIZE_HEADING_DEFAULT: "36px",
  FONT_SIZE_HEADING_SM: "32px",

  FONT_SIZE_SUBHEADING_LG: "18px",
  FONT_SIZE_SUBHEADING_DEFAULT: "16px",
  FONT_SIZE_SUBHEADING_SM: "14px",

  FONT_SIZE_ITEM_TITLE_LG: "16px",
  FONT_SIZE_ITEM_TITLE_DEFAULT: "14px",
  FONT_SIZE_ITEM_TITLE_SM: "12px",

  FONT_SIZE_TEXT_XL: "18px",
  FONT_SIZE_TEXT_LG: "16px",
  FONT_SIZE_TEXT_DEFAULT: "14px",
  FONT_SIZE_TEXT_SM: "12px",
  FONT_SIZE_TEXT_XS: "10px",

  LINE_HEIGHT_LOOSE: 1.6,
  LINE_HEIGHT_DEFAULT: 1.4,
  LINE_HEIGHT_TIGHT: 1.2,

  FONT_WEIGHT_DEFAULT: 400,
  FONT_WEIGHT_MEDIUM: 500,
  FONT_WEIGHT_SEMIBOLD: 600,
  FONT_WEIGHT_BOLD: 700,

  LETTER_SPACING_DEFAULT: "normal"
};
```

You can extend this base theme like so:

```javascript
// ...App.js

keen = {
  ...keen,
  COLOR_BRAND_PRIMARY: "#03aaed"
}

render(
  <ThemeProvider theme={keen}>
    <App />
  </ThemeProvider>, 
  document.querySelector('#app'));
```

You can find more information on style variables [here](https://greymatter-io.github.io/gm-ui-components/?selectedKind=%20Overview%7CStyles&selectedStory=Styles&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fthemes%2Fpanel).

# Documentation

See the [greymatter.io UI Components storybook](https://greymatter-io.github.io/gm-ui-components) for more usage examples and documentation.

# Compatability

* [Full list of browser support matrix](http://browserl.ist/?q=last+2+Chrome+versions%2C+last+2+ChromeAndroid+versions%2C+last+2+Safari+versions%2C+last+2+iOS+versions%2C+last+2+Firefox+versions%2C+last+2+Edge+versions)

# License

MIT


CHANGELOG.md
## [2.1.3](https://github.com/greymatter-io/gm-ui-components/compare/v2.1.2...v2.1.3) (2022-09-30)

### Changed

* Changed all instances of "Grey Matter" that refer to the company to "greymatter.io"
* Changed logo name from "decipher-logo.png" to "greymatter-logo.png"
* Replaced CircleCI pipeline with Buildkite
* Fixed NPM security vulnerabilities
* Fixed theme switching regression


## [2.1.2](https://github.com/greymatter-io/gm-ui-components/compare/v2.1.1...v2.1.2) (2022-07-05)

### Features

* Changed alt text and copyright from Decipher to greymatter.io Inc. ([#636](https://github.com/greymatter-io/gm-ui-components/issues/636)) ([ac02521](https://github.com/greymatter-io/gm-ui-components/commit/ac0252118084e08ba0f3d86c688a873a0565b931))


## [2.1.1](https://github.com/greymatter-io/gm-ui-components/compare/v2.1.0...v2.1.1) (2021-09-03)

### Bug Fixes
* remove console.log statement ([#634](https://github.com/greymatter-io/gm-ui-components/issues/634)) ([51610a9](https://github.com/greymatter-io/gm-ui-components/commit/51610a9f0d7189f45e2154c05faa6d3280455829))


# [2.1.0](https://github.com/greymatter-io/gm-ui-components/compare/v2.0.0...v2.1.0) (2021-09-03)

### Features

* add cross-browser input reset ([#629](https://github.com/greymatter-io/gm-ui-components/issues/629)) ([2c2ed76](https://github.com/greymatter-io/gm-ui-components/commit/2c2ed76272b96ff3d6cb709647497ce1d47f69df))
* auto-generate css vars from theme vars ([#632](https://github.com/greymatter-io/gm-ui-components/issues/632)) ([568462d](https://github.com/greymatter-io/gm-ui-components/commit/568462dcfd10a7dadcd4f749c5b584ed7acb4c9b))
* squished icons remain centered ([#625](https://github.com/greymatter-io/gm-ui-components/issues/625)) ([c4dd3d8](https://github.com/greymatter-io/gm-ui-components/commit/c4dd3d8105d385f21757cf574f5aba3c7e0fdc9d))
* update brand elements ([#624](https://github.com/greymatter-io/gm-ui-components/issues/624)) ([a1b7b68](https://github.com/greymatter-io/gm-ui-components/commit/a1b7b683848511ff48cd087a4858746de5276a66))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.4...v2.0.0) (2020-12-02)

### Breaking Changes

* add inter as a dev dep and set peer deps to major versions only ([#599](https://github.com/greymatter-io/gm-ui-components/issues/599)) ([2f7a8e6](https://github.com/greymatter-io/gm-ui-components/commit/2f7a8e6))
* deprecate nonstandard svg terms in icon API ([#595](https://github.com/greymatter-io/gm-ui-components/issues/595)) ([5e7708f](https://github.com/greymatter-io/gm-ui-components/commit/5e7708f))
* add react-popper and @popperjs/core as peer dependencies ([4af2b73](https://github.com/greymatter-io/gm-ui-components/commit/4af2b73cee4893c596223b9db9ada8f9a2db3ede))
  
### Bug Fixes

* solve cause of buttons having too-solid text color ([#596](https://github.com/greymatter-io/gm-ui-components/issues/596)) ([12d8554](https://github.com/greymatter-io/gm-ui-components/commit/12d8554))
* update footer social links ([#600](https://github.com/greymatter-io/gm-ui-components/issues/600)) ([78d5c45](https://github.com/greymatter-io/gm-ui-components/commit/78d5c45))

### Features

* icons can now be presented in negative ([#533](https://github.com/greymatter-io/gm-ui-components/issues/533)) ([1955398](https://github.com/greymatter-io/gm-ui-components/commit/1955398))
* update svgr ([#612](https://github.com/greymatter-io/gm-ui-components/issues/612)) ([c6db69c](https://github.com/greymatter-io/gm-ui-components/commit/c6db69c))
* better default breadcrumb delimiter (or custom, or none) ([4e426fc](https://github.com/greymatter-io/gm-ui-components/commit/4e426fc87b02c20ef32522fcf05a4fb2fa7e2aca))



<a name="1.4.4"></a>
## [1.4.4](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.3...v1.4.4) (2020-09-23)


### Bug Fixes

* icon displayName ([#590](https://github.com/greymatter-io/gm-ui-components/issues/590)) ([5c9ef2d](https://github.com/greymatter-io/gm-ui-components/commit/5c9ef2d))
* use different inter package ([#589](https://github.com/greymatter-io/gm-ui-components/issues/589)) ([3fcc21d](https://github.com/greymatter-io/gm-ui-components/commit/3fcc21d))
* move inter to peerDependencies ([159ff70](https://github.com/greymatter-io/gm-ui-components/commit/159ff70))



<a name="1.4.3"></a>
## [1.4.3](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.2...v1.4.3) (2020-09-23)

NOTE: this release uses the wrong package for inter fonts, please upgrade to 1.4.4 to take advantage of the new font-family.

### Bug Fixes

* center spinner loading message ([#577](https://github.com/greymatter-io/gm-ui-components/issues/577)) ([67ddd0d](https://github.com/greymatter-io/gm-ui-components/commit/67ddd0d))


### Features

* performance enhancements ([#585](https://github.com/greymatter-io/gm-ui-components/issues/585)) ([150a9f4](https://github.com/greymatter-io/gm-ui-components/commit/150a9f4))
* use Inter for base type ([#574](https://github.com/greymatter-io/gm-ui-components/issues/574)) ([a4eed1a](https://github.com/greymatter-io/gm-ui-components/commit/a4eed1a))



<a name="1.4.2"></a>
## [1.4.2](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.1...v1.4.2) (2020-08-31)


### Features

* added badge component ([#581](https://github.com/greymatter-io/gm-ui-components/issues/581)) ([498ae48](https://github.com/greymatter-io/gm-ui-components/commit/498ae48))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/greymatter-io/gm-ui-components/compare/v1.4.0...v1.4.1) (2020-06-01)

* Added automatic formatting of JS files on git commit
* Removed outdated related projects from README

<a name="1.4.0"></a>
# [1.4.0](https://github.com/greymatter-io/gm-ui-components/compare/v1.3.0...v1.4.0) (2020-04-09)

### Added components

- IconArt
- IconAt
- IconBuilding
- IconBusiness
- IconCalendarEvent
- IconCalendarEventPin
- IconCalendarPlus
- IconCheckcircleIndeterminate
- IconCity
- IconCitySkyline
- IconColumnToggle

src/style/theme.js
import { lighten } from "polished";

export const keen = {
  name: "Keen",

  COLOR_BRAND_PRIMARY: "#00b42b",
  COLOR_BRAND_SECONDARY: "#00b42b",

  COLOR_BACKGROUND_DEFAULT: "hsl(0, 0%, 100%)",
  COLOR_BACKGROUND_TWO: "#FDFDFE",
  COLOR_BACKGROUND_THREE: "hsl(0, 0%, 94%)",

  COLOR_CONTENT_DEFAULT: "hsla(0, 0%, 0%, 0.85)",
  COLOR_CONTENT_CONTRAST: "hsl(0, 0%, 0%)",
  COLOR_CONTENT_MUTED: "hsla(0, 0%, 0%, 0.6)",
  COLOR_CONTENT_NONESSENTIAL: "hsla(0, 0%, 0%, 0.3)",

  COLOR_INTENT_HIGHLIGHT: "#00b42b",
  COLOR_INTENT_SUCCESS: "#00b42b",
  COLOR_INTENT_DANGER: "#D83D22", // WCAG AA+
  COLOR_INTENT_WARNING: "#F7CD45", // WCAG AA+
  COLOR_INTENT_INFO: "#1E6DF6", // WCAG AA+

  COLOR_KEYLINE_DEFAULT: "hsla(0, 0%, 0%, 0.08)",
  COLOR_KEYLINE_SOLID: "hsl(0, 0%, 92%)",

  OPACITY_FULL: "1",
  OPACITY_LIGHT: "0.7",
  OPACITY_LIGHTER: "0.5",
  OPACITY_LIGHTEST: "0.15",

  // LAYOUT
  SPACING_BASE: 8,

  ZINDEX_ABYSS: "-9999",
  ZINDEX_FLOOR: "1",
  ZINDEX_DROPDOWN: "1010",
  ZINDEX_STICKY: "1020",
  ZINDEX_FIXED: "1030",
  ZINDEX_SCRIM: "1040",
  ZINDEX_MODAL: "1050",
  ZINDEX_POPOVER: "1060",
  ZINDEX_TOOLTIP: "1070",

  CORNER_RADIUS_SHARP: "2px", // spacingScale(0.5)
  CORNER_RADIUS_INPUT: "4px", // spacingScale(1)
  CORNER_RADIUS_CARD_SM: "4px",
  CORNER_RADIUS_CARD_DEFAULT: "6px",
  CORNER_RADIUS_CARD_LG: "8px",
  CORNER_RADIUS_MAX: "90000px",

  // TYPOGRAPHY
  FONT_STACK_DEFAULT: `Inter var, Inter, -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif`,
  FONT_STACK_BRAND: `"Metropolis", "Helvetica Neue", Arial, sans-serif`,
  FONT_STACK_CODE: `"SFMono-Regular", Consolas, "Liberation Mono", Menlo, Courier, monospace`,

  FONT_SIZE_PAGE_TITLE: "40px",

  FONT_SIZE_HEADING_LG: "38px",
  FONT_SIZE_HEADING_DEFAULT: "36px",
  FONT_SIZE_HEADING_SM: "32px",

  FONT_SIZE_SUBHEADING_LG: "18px",
  FONT_SIZE_SUBHEADING_DEFAULT: "16px",
  FONT_SIZE_SUBHEADING_SM: "14px",

  FONT_SIZE_ITEM_TITLE_LG: "16px",
  FONT_SIZE_ITEM_TITLE_DEFAULT: "14px",
  FONT_SIZE_ITEM_TITLE_SM: "12px",

  FONT_SIZE_TEXT_XL: "18px",
  FONT_SIZE_TEXT_LG: "16px",
  FONT_SIZE_TEXT_DEFAULT: "14px",
  FONT_SIZE_TEXT_SM: "12px",
  FONT_SIZE_TEXT_XS: "10px",

  LINE_HEIGHT_LOOSE: 1.6,
  LINE_HEIGHT_DEFAULT: 1.4,
  LINE_HEIGHT_TIGHT: 1.2,

  FONT_WEIGHT_DEFAULT: 400,
  FONT_WEIGHT_MEDIUM: 500,
  FONT_WEIGHT_SEMIBOLD: 600,
  FONT_WEIGHT_BOLD: 700,

  LETTER_SPACING_DEFAULT: "normal"
};

const LIGHTENING_FACTOR = 0.05;

export const keenDark = {
  ...keen,

  name: "KeenDark",

  COLOR_BACKGROUND_DEFAULT: "#111",
  COLOR_BACKGROUND_TWO: "#181818",
  COLOR_BACKGROUND_THREE: "#242424",

  COLOR_CONTENT_DEFAULT: "hsla(0, 0%, 100%, 0.8)",
  COLOR_CONTENT_CONTRAST: "hsl(0, 0%, 100%)",
  COLOR_CONTENT_MUTED: "hsla(0, 0%, 100%, 0.6)",
  COLOR_CONTENT_NONESSENTIAL: "hsla(0, 0%, 100%, 0.5)",

  COLOR_INTENT_DANGER: lighten(LIGHTENING_FACTOR, keen.COLOR_INTENT_DANGER), // WCAG AA+
  COLOR_INTENT_WARNING: lighten(LIGHTENING_FACTOR, keen.COLOR_INTENT_WARNING), // WCAG AA+
  COLOR_INTENT_INFO: lighten(LIGHTENING_FACTOR, keen.COLOR_INTENT_INFO), // WCAG AA+

  COLOR_KEYLINE_DEFAULT: "hsla(0, 0%, 100%, 0.1)",
  COLOR_KEYLINE_SOLID: "hsla(0, 0%, 14%, 1)",

  LINE_HEIGHT_LOOSE: 1.7,
  LINE_HEIGHT_DEFAULT: 1.5,
  LINE_HEIGHT_TIGHT: 1.3,

  FONT_WEIGHT_DEFAULT: 500,
  FONT_WEIGHT_MEDIUM: 600,
  FONT_WEIGHT_SEMIBOLD: 700,
  FONT_WEIGHT_BOLD: 800,

  LETTER_SPACING_DEFAULT: "0.025ch"
};
